### PR TITLE
Processing Data in Models

### PR DESCRIPTION
**Pull request by DJ-MrJay**
In this branch, I:
- Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
- Set up associations between models.
- Use models to insert data.
- Create custom methods.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id'
+  has_many :likes, foreign_key: 'author_id'
+  has_many :comments, through: :posts
+
+  validates :name, presence: true
```
Comment (zainsadaqat) [inline app/models/user.rb:6]: The user model must have a method that returns the 3 most recent posts as part of the project requirement. Kindly implement that method. 👍

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comments
+  has_many :likes
+
+  validates :title, presence: true
+  validates :text, presence: true
```
Comment (zainsadaqat) [inline app/models/post.rb:7]: The post model must have a method that updates the posts counter for a user as it's part of the project requirement. Kindly implement that 👍

Comment (zainsadaqat) [inline app/models/post.rb:7]: The Post model must have a method that returns the 5 most recent comments for a given post as it's part of project requirement. Kindly implement that 👍

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
+
+  validates :text, presence: true
```
Comment (zainsadaqat) [inline app/models/comment.rb:5]: The comment model must have a method that updates the comments counter for a post as it's part of the requirement. Kindly implement that 👍

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  belongs_to :post
```
Comment (zainsadaqat) [inline app/models/like.rb:3]: Like model must have a method that updates the likes counter for a post as it's a project requirement. Kindly implement that 👍

**Review by zainsadaqat — CHANGES_REQUESTED** [review]
# Required Changes 🔁
## Hello  @DJ-MrJay 👋 🙂

You're doing a great job on your project. We're proud of you. Good job so far! 

It's OK to have bugs in your code. Please, see this code review as a chance for you to learn practical coding skills and learn new tips that can improve your code quality and help you get a better job. 

There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
  ✅  Correct use of GitHub Flow
  ✅  Descriptive Pull Request
  ✅  No linter errors

## Required Changes ♻️

- [ ] 



## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better._

Cheers and Happy coding! 👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+  after_save :update_post_comments_counter
+
+  private
+
+  def update_post_comments_counter
+    post.update(comments_count: post.comments.count)
+  end
```
Comment (Qoosim) [inline app/models/comment.rb:13]: - [OPTIONAL] Nice job implementing required changes from the previous reviewer. I see below snippet more clean. You can check it out.
```
      def update_comments_counter
          post.increment!(:comments_counter)
      end

```

You can apply the same to other models where necessary if you like it :smile:

**Review by Qoosim — APPROVED** [review]
#  *Status: Approved* :tada: :heavy_check_mark: 

Hi @DJ-MrJay, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
-  No linter errors. :heavy_check_mark: 
-  Good README file. :heavy_check_mark: 
-  Correct Gitflow. :heavy_check_mark: 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by prolajumokeoni — COMMENTED** [review]
Hi @DJ-MrJay 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
## Highlights
-  The associations between models are set up correctly 👏🏿


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**